cinema/views.py

```diff
@@ -31,28 +46,90 @@ class CinemaHallViewSet(viewsets.ModelViewSet):
 
 
 class MovieViewSet(viewsets.ModelViewSet):
-    queryset = Movie.objects.all()
+    queryset = Movie.objects.prefetch_related("genres", "actors")
     serializer_class = MovieSerializer
 
     def get_serializer_class(self):
         if self.action == "list":
             return MovieListSerializer
-
-        if self.action == "retrieve":
+        elif self.action == "retrieve":
             return MovieDetailSerializer
+        return super().get_serializer_class()
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+
+        genres_param = self.request.query_params.get("genres")
+        actors_param = self.request.query_params.get("actors")
+        title_param = self.request.query_params.get("title")
+
+        if self.action == "list":
+            if genres_param:
+                genre_ids = parse_query_to_ints(genres_param)
+                queryset = queryset.filter(genres__id__in=genre_ids)
+
+            if actors_param:
+                actor_ids = parse_query_to_ints(actors_param)
+                queryset = queryset.filter(actors__id__in=actor_ids)
+
+            if title_param:
+                queryset = queryset.filter(title__icontains=title_param)
 
-        return MovieSerializer
+        return queryset.distinct()
 
 
 class MovieSessionViewSet(viewsets.ModelViewSet):
-    queryset = MovieSession.objects.all()
+    queryset = (
+        MovieSession.objects.select_related
+        ("movie", "cinema_hall").prefetch_related("tickets").all()
```

Comment:
you don't need to prefetch tickets on every action and you don't need all() in the end